### ISSUE TITLE
Support merge trains (any platform)

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#6011]
**What would you like Renovate to be able to do?**

Similar to the idea in #5573 but to be implemented completely within Renovate's own logic.

**Describe the solution you'd like**

Something that means multiple PRs don't need to be rebased after each merge.

**Describe alternatives you've considered**

Having a "roll up" PR that adds all passing updates (i.e. test them individually, then add to a single group PR after).

**Additional context**

https://github.com/renovatebot/config-help/issues/610

**Comment by HonkingGoose** [comment]
Have you seen GitHub's auto-merge functionality? (It's in beta for now).

> If you enable auto-merge for a pull request, the pull request will merge automatically when all required reviews are met and status checks have passed. Auto-merge prevents you from waiting around for requirements to be met, so you can move on to other tasks.

> Before you can use auto-merge with a pull request, auto-merge must be enabled for the repository. 

Maybe GitHub will let you use this feature through an API later.
Or maybe you can find some other way to leverage the functionality once it's had a stable release.

Just wanted to let you know that this is a thing. 😄 

You can read the [GitHub docs, Automatically merging a pull request](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/automatically-merging-a-pull-request) for more information.

**Comment by rarkins** [comment]
Thanks, but I don't think this will be helpful here. We don't have a problem with merging - we have a problem with lots of rebasing, which GitHub automerge wouldn't help.

**Comment by HonkingGoose** [comment]
For future reference: the alternative idea of "rollup" updates has its own issue (#4404).

**Comment by HonkingGoose** [comment]
@rarkins GitHub is starting work on something similar to GitLab's Merge train, they're calling it a "Pull Request Merge Queue".

Relevant quote from their issue:  [^github-merge-queue]

> ## How will it work?
> 
> A developer can add their pull request to the merge queue when:
> 1. Merge queue is enabled on a branch via a new branch protection setting, and...
> 1. All merge requirements have been met (e.g. including required reviewer and status check)
> 
> Once in the queue, the pull request's changes will be:
> 
> 1. validated to ensure it passes all required checks before merging
> 1. merged with the latest version of the branch along with any pull requests ahead of it in the queue

[^github-merge-queue]: https://github.com/github/roadmap/issues/272

**Comment by HonkingGoose** [comment]
> I'm wondering if it's more about "keep rebasing until merged" rather than merge trains - which intentionally rebase themselves on top of the PR they trail in the queue (it's more efficient)

I just checked the GitLab docs on merge trains so I'm on the same page as you. [^merge-train-example]

> ## Merge train example
> 
> Three merge requests (`A`, `B` and `C`) are added to a merge train in order, which creates three merged results pipelines that run in parallel:
> 
> 1. The first pipeline runs on the changes from `A` combined with the target branch.
> 2. The second pipeline runs on the changes from `A` and `B` combined with the target branch.
> 3. The third pipeline runs on the changes from `A`, `B`, and `C` combined with the target branch. 
> 
> If the pipeline for `B` fails, it is removed from the train. The pipeline for `C` restarts with the `A` and `C` changes, but without the `B` changes.
> 
> If `A` then completes successfully, it merges into the target branch, and `C` continues to run. If more merge requests are added to the train, they now include the `A` changes that are included in the target branch, and the `C` changes that are from the merge request already in the train.

Wow! I'm really impressed with the smart behavior that GitLab managed to build! 🤯 

Maybe GitHub is thinking of something similar, it's a bit hidden in this quote:

> Once in the queue, the pull request's changes will be:
> 
> 1. validated to ensure it passes all required checks before merging
> 1. merged with the latest version of the branch **along with any pull requests ahead of it in the queue**

The bolded part sounds a bit similar to GitLabs merge train?

In any case, even if GitHub "only" builds a "brute-force" mechanism that auto-updates the PR branch with `main` that would be a big help compared to the current behavior. 😄 

[^merge-train-example]: https://docs.gitlab.com/ee/ci/pipelines/merge_trains.html#merge-train-example

**Comment by rarkins** [comment]
Yes, that quote does sound like they're planning something similar

**Comment by mikaelkolkinn** [comment]
Just letting you know that some changes seem to be necessary in order for Renovate to merge using GitHubs new Merge Queue. We have enabled it on one of our repositories recently, and since then no Renovate PRs have been merged autmatically for that repository.

**Comment by rarkins** [comment]
Are you attempting to manually add a Renovate PR to your branch's merge queue as per the description in https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/merging-a-pull-request-with-a-merge-queue ?

**Comment by mikaelkolkinn** [comment]
This might differ from one beta tester to another beta tester if GitHub is doing A/B testing, but in our case it seems necessary to **require** the use of merge queues in order to be able to use it at all.
It is with the mandatory merge queues that Renovate cannot automerge. Manually adding a Renovate PR to the merge queue using the "Merge when ready" button does work, just like a regular merge.

With mandatory merge queues enabled:
![image](https://user-images.githubusercontent.com/3751736/165707759-02324b04-7737-4221-943f-1daf8f4ad010.png)
![image](https://user-images.githubusercontent.com/3751736/165706700-ce4616dd-48cb-44ff-8fef-a0ba7ffa04d9.png)

Without mandatory merge queues enabled:
![image](https://user-images.githubusercontent.com/3751736/165706886-000d883f-145d-4010-bcf1-f9bb35c7ef35.png)
![image](https://user-images.githubusercontent.com/3751736/165706944-b25f7d7a-feda-4f74-946f-aeaac87e55b4.png)

**Comment by rarkins** [comment]
It sounds like the current approach of merge queues is potentially incompatible with the concept of automerging, unless we can access those options via API

**Comment by adamstrawson** [comment]
Is this a feature that is planned to be implemented? We're having to manually merge Renovate PR's at the moment as we're using GitHub merge queues, but it'll be nice to be able to auto-merge as before so that it's not a chore our engineers are having to maintain.

**Comment by rarkins** [comment]
Nothing in `status:requirements` is planned, because it's not ready to start. I looked in the above history and don't think there's any consensus on what needs to be done

**Comment by HonkingGoose** [comment]
Here's the new upstream issue for general release of the GitHub merge queue:

- https://github.com/github/roadmap/issues/370

The [GitHub Docs, Merging a pull request with a merge queue](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/merging-a-pull-request-with-a-merge-queue) have some more information about how it works in the GitHub interface. But that's probably not much help for the maintainers of Renovate.

The feature is still in beta, per the GitHub docs:

> **Note:** The pull request merge queue feature is currently in limited public beta and subject to change.

**Comment by rarkins** [comment]
Functionally, is this a new type of "automerge" where Renovate adds it to a merge queue once requirements are met, instead of enabling automerge at PR creation or automerging itself?

Technically, does the API for this exist yet?

**Comment by HonkingGoose** [comment]
I don't know if there's an API yet, but there's a new webhook [^webhook] event and GitHub Actions workflow trigger [^trigger] called `merge_group`. Quote from the GitHub Blog: [^gh-blog]

> A new webhook event and GitHub Actions workflow trigger (`merge_group`) makes it easier to run required status checks on merge groups created by merge queue. A merge group includes the changes from one or more pull requests and must pass the status checks required by the target branch.

I recommend you read all three links, and see if it's what you need. 😉 

[^webhook]: https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#merge_group

[^trigger]: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#merge_group

[^gh-blog]: https://github.blog/changelog/2022-08-18-merge-group-webhook-event-and-github-actions-workflow-trigger/

**Comment by HonkingGoose** [comment]
There's no way to use Renovate + GitHub Merge Queue + Automerge (squash). The Merge Queue feature is still in limited public beta, and we didn't get in the beta.

I think we need to have access to the Merge Queue feature first. That way we can test the feature and see how to integrate it into Renovate bot.

**Comment by kaelig** [comment]
Thank you for the context, I'll bring it up to GitHub in an upcoming discussion about the Netlify Deploy Preview <> GitHub Merge Queue developer workflow.

If you'd like, I can ask that they enroll you to the beta?

**Comment by rarkins** [comment]
Yes please do - best would be this open source org "renovatebot" but if they insist that it should be an enterprise customer only then my and Renovate's parent company "mend" would still be useful for me to test with if included.

**Comment by vchernin** [comment]
The github merge queue feature [is in public beta now](https://github.blog/changelog/2023-02-08-pull-request-merge-queue-public-beta/) and I have been playing around with it. The [merge_group](https://github.com/community/community/discussions/46757#discussioncomment-4909983) event is currently broken but otherwise it seems usable.

**Comment by rarkins** [comment]
Does it support an API for Renovate to be able to add its PRs to the queue?

**Comment by vchernin** [comment]
As far as I can tell there isn't a REST api, but there is a github graphql mutation [`enablePullRequestAutoMerge`](https://docs.github.com/en/graphql/reference/mutations#enablepullrequestautomerge). Tried it on my repo using https://docs.github.com/en/graphql/overview/explorer and it works as expected.

```graphql
query get-pr-id {
    repository(owner: "$REPO_ORG", name: "$REPO_NAME") {
        pullRequest(number: PR_NUMBER) {
                  id
              }
        } 
}

mutation add-to-queue {
    enablePullRequestAutoMerge(input: {pullRequestId: "id-from-previous-query", mergeMethod: MERGE}) {
        clientMutationId
         }
}
```

**Comment by vchernin** [comment]
Actually, Renovate can already add to GitHub merge queues. It is necessary to enable auto merge in repo settings (even though a merge queue is being used). Renovate is already making the right API call: https://github.com/renovatebot/renovate/blob/bde951257a79737351859d1c08db0ad214e61494/lib/modules/platform/github/index.ts#L1455-L1458

So it ends up looking like so:

![DB11D45C-FF54-4D95-A10E-54088367BA71](https://user-images.githubusercontent.com/38842733/217890493-a26b30c2-fa73-4063-98de-6514257326f7.jpeg)

This is using platform auto merge:

```json
{
    "automerge": true,
    "automergeType": "pr",
    "platformAutomerge": true
}
```

**Comment by HonkingGoose** [comment]
> BTW I think once it's technically feasible, we should break GitHub Merge trains into its own issue. This one was intended originally as a "generic" way to achieve the same without native platform support

Let's discuss the _GitHub_ Merge Queue in this issue from now on:

- #20321
